lib/puma/client.rb

```diff
@@ -300,6 +300,48 @@ def finish(timeout)
       @to_io.wait_readable(timeout) || timeout! until try_to_finish
     end
 
+    # Wraps `@parser.execute` and adds meaningful error messages
+    # @return [Integer] bytes of buffer read by parser
+    #
+    def parser_execute
+      @parser.execute(@env, @buffer, @parsed_bytes)
+    rescue => e
+      @env[HTTP_CONNECTION] = 'close'
+      req, _ = @buffer.split "\r\n\r\n"
+      request_line, headers = req.split "\r\n", 2
+      if HttpParserError === e
+        if e.message.include?('non-SSL')
```

Comment:
Maybe guard here too instead of test

Reply:
I added another commit cleaning up `parser_execute`.